src/nlp_expr.jl

```diff
@@ -0,0 +1,871 @@
+            new_ret = GenericNonlinearExpr{V}(arg.head, Any[])
+            push!(parent.args, new_ret)
+            for child in reverse(arg.args)
+                push!(stack, (new_ret, child))
+            end
+        elseif arg isa Number
+            push!(parent.args, arg)
+        else
+            push!(parent.args, jump_function(model, arg))
+        end
+    end
+    return ret
+end
+
+function jump_function_type(
+    model::GenericModel,
+    ::Type{<:MOI.ScalarNonlinearFunction},
+)
+    return GenericNonlinearExpr{variable_ref_type(typeof(model))}
+end
+
+moi_function_type(::Type{<:GenericNonlinearExpr}) = MOI.ScalarNonlinearFunction
+
+function constraint_object(c::NonlinearConstraintRef)
+    nlp = nonlinear_model(c.model)
+    data = nlp.constraints[index(c)]
+    return ScalarConstraint(jump_function(c.model, data.expression), data.set)
+end
+
+function jump_function(model::GenericModel, expr::MOI.Nonlinear.Expression)
+    V = variable_ref_type(typeof(model))
+    nlp = nonlinear_model(model)
+    parsed = Vector{Any}(undef, length(expr.nodes))
+    adj = MOI.Nonlinear.adjacency_matrix(expr.nodes)
+    rowvals = SparseArrays.rowvals(adj)
+    for i in length(expr.nodes):-1:1
+        node = expr.nodes[i]
+        parsed[i] = if node.type == MOI.Nonlinear.NODE_CALL_UNIVARIATE
+            GenericNonlinearExpr{V}(
+                nlp.operators.univariate_operators[node.index],
+                parsed[rowvals[SparseArrays.nzrange(adj, i)[1]]],
+            )
+        elseif node.type == MOI.Nonlinear.NODE_CALL_MULTIVARIATE
+            GenericNonlinearExpr{V}(
+                nlp.operators.multivariate_operators[node.index],
+                Any[parsed[rowvals[j]] for j in SparseArrays.nzrange(adj, i)],
+            )
+        elseif node.type == MOI.Nonlinear.NODE_MOI_VARIABLE
+            V(model, MOI.VariableIndex(node.index))
+        elseif node.type == MOI.Nonlinear.NODE_PARAMETER
+            NonlinearParameter(model, node.index)
+        elseif node.type == MOI.Nonlinear.NODE_SUBEXPRESSION
+            NonlinearExpression(model, node.index)
+        elseif node.type == MOI.Nonlinear.NODE_VALUE
+            expr.values[node.index]
+        else
+            # node.type == MOI.Nonlinear.NODE_COMPARISON
+            # node.type == MOI.Nonlinear.NODE_LOGIC
+            error("Unsupported node")
+        end
+    end
+    return parsed[1]
+end
+
+function value(f::Function, expr::GenericNonlinearExpr)
+    return _evaluate_expr(MOI.Nonlinear.OperatorRegistry(), f, expr)
+end
+
+function value(a::GenericNonlinearExpr; result::Int = 1)
+    return value(a) do x
+        return value(x; result = result)
+    end
+end
+
+function _evaluate_expr(
+    ::MOI.Nonlinear.OperatorRegistry,
+    f::Function,
+    expr::AbstractJuMPScalar,
+)
+    return value(f, expr)
+end
+
+function _evaluate_expr(
+    ::MOI.Nonlinear.OperatorRegistry,
+    ::Function,
+    expr::Real,
+)
+    return convert(Float64, expr)
```

Comment:
Is forcing `Float64` here compatible with generic typing?

Reply:
The nonlinear interface is still `Float64` only

Reply:
Is that fact documented somewhere?

Reply:
We could try seeing what the impact of relaxing this is. the main problem is not that JuMP requires `Float64`, but that `MOI.Nonlinear` does.

Reply:
This is one problem:
```Julia
julia> @which MOI.Nonlinear.sin
NaNMath

julia> MOI.Nonlinear.sin(big"1.0")
ERROR: StackOverflowError:
Stacktrace:
 [1] sin(x::BigFloat) (repeats 79984 times)
   @ NaNMath ~/.julia/packages/NaNMath/ceWIc/src/NaNMath.jl:11
```
NaNMath doesn't have the right fallbacks for `BigFloat`.